File: app/src/main/java/com/example/client/screens/wallet/activity/RechargeActivity.java
package com.example.client.screens.wallet.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.example.client.R;
import com.example.client.app.Constrants;
import com.example.client.dialog.PrimaryDialog;
import com.example.client.models.message.MessageModel;
import com.example.client.models.profile.ProfileModel;
import com.example.client.models.subject.SubjectModel;
import com.example.client.models.transaction.TransactionModel;
import com.example.client.screens.wallet.fragment.IWalletView;
import com.example.client.screens.wallet.present.WalletPresent;

import java.util.List;

public class RechargeActivity extends AppCompatActivity implements View.OnClickListener, IWalletView {
    private ImageView back;
    private TextView section_1,section_2,section_5;
    private TextView request;
    private ProgressBar progressBar;
    private WalletPresent wPresent;
    private PrimaryDialog dialog;
    private Double amount;
    private ProfileModel user;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recharge);
        back = findViewById(R.id.back);
        section_1 = findViewById(R.id.section_1);
        section_2 = findViewById(R.id.section_2);
        section_5 = findViewById(R.id.section_5);
        request = findViewById(R.id.request);
        progressBar = findViewById(R.id.progress_bar);

        dialog = new PrimaryDialog();
        dialog.getInstance(this);
        wPresent = new WalletPresent(this);

        back.setOnClickListener(this);
        request.setOnClickListener(this);
        section_1.setOnClickListener(this);
        section_2.setOnClickListener(this);
        section_5.setOnClickListener(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        wPresent.onShowInfoUser();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.back:
                finish();
                onBackPressed();
                break;
            case R.id.request:
                wPresent.onRequestRecharge(user.getId(),amount);
                break;
            case R.id.section_1:
                amount = Constrants.SECTION_1;
                section_1.setBackgroundResource(R.drawable.bg_btn);
                section_1.setTextColor(Color.WHITE);
                section_2.setBackgroundResource(R.drawable.border_item);
                section_2.setTextColor(Color.BLACK);
                section_5.setBackgroundResource(R.drawable.border_item);
                section_5.setTextColor(Color.BLACK);
                request.setBackgroundResource(R.drawable.bg_btn);
                request.setEnabled(true);
                break;
            case R.id.section_2:
                amount = Constrants.SECTION_2;
                section_2.setBackgroundResource(R.drawable.bg_btn);
                section_2.setTextColor(Color.WHITE);
                section_1.setBackgroundResource(R.drawable.border_item);
                section_1.setTextColor(Color.BLACK);
                section_5.setBackgroundResource(R.drawable.border_item);
                section_5.setTextColor(Color.BLACK);
                request.setBackgroundResource(R.drawable.bg_btn);
                request.setEnabled(true);
                break;
            case R.id.section_5:
                amount = Constrants.SECTION_5;
                section_5.setBackgroundResource(R.drawable.bg_btn);
                section_5.setTextColor(Color.WHITE);
                section_2.setBackgroundResource(R.drawable.border_item);
                section_2.setTextColor(Color.BLACK);
                section_1.setBackgroundResource(R.drawable.border_item);
                section_1.setTextColor(Color.BLACK);
                request.setBackgroundResource(R.drawable.bg_btn);
                request.setEnabled(true);
                break;
        }
    }


    @Override
    public void showInfoUser(ProfileModel user) {
        this.user = user;
    }

    @Override
    public void showTransaction(List<TransactionModel> items, String method) {

    }

    @Override
    public void showTransactionByOrderCode(TransactionModel item) {

    }

    @Override
    public void getSubject(SubjectModel item) {

    }

    @Override
    public void deleteRecharge(MessageModel message) {

    }

    @Override
    public void requestRecharge(MessageModel message) {
        if(this!=null){

        }
        if(message.isStatus()){
            dialog.setDescription("Yêu cầu nạp tiền đã được gửi, vui lòng lên văn phòng để nạp tiền");
            dialog.hideBtnCancel();
            dialog.show();
            dialog.setOKListener(()->{
                finish();
                onBackPressed();
            });
        }
        else {
            switch (message.getCode()){
                case Constrants.ErrorCode.ERROR_1001:
                    dialog.setDescription(getString(R.string.err_code_1001));
                    break;
            }
            dialog.setOKListener(()->{});
            dialog.hideBtnCancel();
            dialog.show();
        }
    }

    @Override
    public void refreshUserActive(MessageModel message) {

    }

    @Override
    public void showLoading() {
        progressBar.setVisibility(View.VISIBLE);
        request.setBackgroundResource(R.drawable.bg_btn_disable);
        request.setText("");
        request.setEnabled(false);
    }

    @Override
    public void hideLoading() {
        progressBar.setVisibility(View.GONE);
        request.setBackgroundResource(R.drawable.bg_btn);
        request.setText("Lưu thay đổi");
        request.setEnabled(true);
    }
}

File: app/src/main/java/com/example/client/screens/wallet/recharge/RechargeActivity.kt
package com.example.client.screens.wallet.recharge

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.View
import com.example.client.R
import com.example.client.app.Momo
import com.example.client.base.BaseActivityMVP
import com.example.client.dialog.PrimaryDialog
import com.example.client.screens.profile.navigate.NavigatorProfile
import com.example.client.screens.wallet.recharge.present.IRechargePresent
import com.example.client.screens.wallet.recharge.present.RechargePresent
import com.example.client.utils.MomoCallBack
import kotlinx.android.synthetic.main.activity_recharge.*
import java.text.NumberFormat
import java.util.*

class RechargeActivity : BaseActivityMVP<IRechargePresent>(), View.OnClickListener, IRechargeView {

    companion object {
        @JvmStatic
        fun newInstance(from: Activity) = Intent(from, RechargeActivity::class.java)
    }

    override val presenter: IRechargePresent
        get() = RechargePresent(this)

    private var amount = 0.0
    private val sections by lazy { arrayListOf(1000000.0, 2000000.0, 5000000.0) }
    private val formatVND by lazy { NumberFormat.getCurrencyInstance(Locale("vi", "VN")) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recharge)
    }

    override fun bindComponent() {
        tv_section_first.text = formatVND.format(sections[0])
        tv_section_second.text = formatVND.format(sections[1])
        tv_section_thirst.text = formatVND.format(sections[2])
    }

    override fun bindEvent() {
        tv_section_first.setOnClickListener(this)
        tv_section_second.setOnClickListener(this)
        tv_section_thirst.setOnClickListener(this)
        imv_back.setOnClickListener(this)
        tv_submit.setOnClickListener(this)
    }

    override fun showToast() {
        showToastMessage("Thành công")
    }

    override fun showErrorMessage(errorMessage: Int) {
        showDialogErrorMessage(getString(errorMessage))
    }

    override fun rechargeSuccess() {
        PrimaryDialog({
            finish()
        }, { })
                .setDescription(getString(R.string.recharge_success))
                .show(supportFragmentManager)
    }


    override fun showLoading() {
        rll_loading.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        rll_loading.visibility = View.GONE

    }

    override fun onBackPress() {
        onBackPressed()
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.imv_back -> onBackPressed()
            R.id.tv_submit -> Momo.newInstance().requestMoMo(this, amount.toInt())
            R.id.tv_section_first -> selectSection(0)
            R.id.tv_section_second -> selectSection(1)
            R.id.tv_section_thirst -> selectSection(2)
        }
    }

    private fun selectSection(index: Int) {
        when (index) {
            0 -> {
                tv_section_first.setBackgroundResource(R.drawable.bg_btn)
                tv_section_first.setTextColor(Color.WHITE)
                tv_section_second.setBackgroundResource(R.drawable.border_item)
                tv_section_second.setTextColor(Color.BLACK)
                tv_section_thirst.setBackgroundResource(R.drawable.border_item)
                tv_section_thirst.setTextColor(Color.BLACK)
            }
            1 -> {
                tv_section_first.setBackgroundResource(R.drawable.border_item)
                tv_section_first.setTextColor(Color.BLACK)
                tv_section_second.setBackgroundResource(R.drawable.bg_btn)
                tv_section_second.setTextColor(Color.WHITE)
                tv_section_thirst.setBackgroundResource(R.drawable.border_item)
                tv_section_thirst.setTextColor(Color.BLACK)
            }
            2 -> {
                tv_section_first.setBackgroundResource(R.drawable.border_item)
                tv_section_first.setTextColor(Color.BLACK)
                tv_section_second.setBackgroundResource(R.drawable.border_item)
                tv_section_second.setTextColor(Color.BLACK)
                tv_section_thirst.setBackgroundResource(R.drawable.bg_btn)
                tv_section_thirst.setTextColor(Color.WHITE)
            }
        }
        tv_submit.setBackgroundResource(R.drawable.bg_btn)
        tv_submit.isEnabled = true
        amount = sections[index]
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK) Momo.newInstance().callbackResult(requestCode, data, object : MomoCallBack {
            override fun onSuccess(data: String, userPhone: String) {
                presenter.createRequest(amount, userPhone, data)
            }

            override fun onError() {
                showDialogErrorMessage(getString(R.string.err_code_1001))
            }

        })
    }


//    fun toHash(json: String):String  {
//        var encoded = ""
//        try {
//            val publicBytes: ByteArray = Base64.decode(Constants.MoMoConstants.PUBLIC_KEY, Base64.DEFAULT)
//            val keySpec = X509EncodedKeySpec(publicBytes)
//            val keyFactory = KeyFactory.getInstance("RSA")
//            val pubKey = keyFactory.generatePublic(keySpec)
//            val cipher: Cipher = Cipher.getInstance("RSA/ECB/PKCS1PADDING") //or try with "RSA"
//            cipher.init(Cipher.ENCRYPT_MODE, pubKey)
//            encoded = Base64.encodeToString(cipher.doFinal(json.toByteArray()), Base64.DEFAULT)
//        } catch (e: Exception) {
//            e.printStackTrace()
//        }
//        return encoded
//    }
}